File: typescript/src/mediator/mediator.test.ts
import { describe, expect, it } from "vitest";
import { ConcreteColleague1 } from "./concrete-colleague1.ts";
import { ConcreteColleague2 } from "./concrete-colleague2.ts";
import { ConcreteMediator } from "./concrete-mediator.ts";
import type { Mediator } from "./mediator.ts";

describe("Mediator", () => {
  it("routes messages to the other colleague in both directions", () => {
    const mediator = new ConcreteMediator();
    const one = new ConcreteColleague1(mediator);
    const two = new ConcreteColleague2(mediator);
    mediator.setColleague1(one);
    mediator.setColleague2(two);
    one.send("a");
    one.send("b");
    two.send("c");
    expect(two.received()).toEqual(["a", "b"]);
    expect(one.received()).toEqual(["c"]);
  });

  it("a message with no partner registered is dropped", () => {
    const mediator = new ConcreteMediator();
    const lonely = new ConcreteColleague1(mediator);
    mediator.setColleague1(lonely);
    lonely.send("anyone?");
    expect(lonely.received()).toEqual([]);
  });

  it("colleagues talk only to the mediator", () => {
    const seen: string[] = [];
    const recording: Mediator = { notify: (sender, message) => void seen.push(`${sender.name()}:${message}`) };
    const one = new ConcreteColleague1(recording);
    const two = new ConcreteColleague2(recording);
    one.send("x");
    two.send("y");
    expect(seen).toEqual(["ConcreteColleague1:x", "ConcreteColleague2:y"]);
    expect(one.received()).toEqual([]);
  });
});

File: java/src/test/java/com/penapereira/patterns/mediator/MediatorTest.java
package com.penapereira.patterns.mediator;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.Test;

class MediatorTest {

    @Test
    void routesMessagesToTheOtherColleagueInBothDirections() {
        ConcreteMediator mediator = new ConcreteMediator();
        Colleague one = new ConcreteColleague1(mediator);
        Colleague two = new ConcreteColleague2(mediator);
        mediator.setColleague1(one);
        mediator.setColleague2(two);
        one.send("a");
        one.send("b");
        two.send("c");
        assertEquals(List.of("a", "b"), two.received());
        assertEquals(List.of("c"), one.received());
    }

    @Test
    void colleaguesTalkOnlyToTheMediator() {
        List<String> seen = new ArrayList<>();
        Mediator recording = (sender, message) -> seen.add(sender.name() + ":" + message);
        Colleague one = new ConcreteColleague1(recording);
        Colleague two = new ConcreteColleague2(recording);
        one.send("x");
        two.send("y");
        assertEquals(List.of("ConcreteColleague1:x", "ConcreteColleague2:y"), seen);
        assertEquals(List.of(), one.received());
        assertEquals(List.of(), two.received());
    }
}
